runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/ProcessRemoteBundleOperation.java

```diff
@@ -100,6 +218,80 @@ public void finish() throws Exception {
+
+    TimerInternals.TimerData timerData = stepContext.getNextFiredTimer(GlobalWindow.Coder.INSTANCE);
+    while (timerData != null) {
+      StateNamespaces.WindowNamespace windowNamespace =
+          (StateNamespaces.WindowNamespace) timerData.getNamespace();
+      BoundedWindow window = windowNamespace.getWindow();
+
+      WindowedValue<KV<Object, Timer>> timerValue =
+          WindowedValue.of(
+              KV.of(
+                  timerIdToKey.get(timerData.getTimerId()),
+                  Timer.of(timerData.getTimestamp(), timerIdToPayload.get(timerData.getTimerId()))),
+              timerData.getTimestamp(),
+              Collections.singleton(window),
+              PaneInfo.NO_FIRING);
+
+      String timerInputPCollection =
+          timerIdToTimerSpecMap.get(timerData.getTimerId()).inputCollectionId();
+
+      // TODO(BEAM-6274): check this for performance considerations.
+      // Timers are allowed to A) fire other timers, and B) fire themselves. To implement this,
+      // we send the timer to the SDK Harness, wait for any elements, then check if any additional
+      // timers were fired.
+      RemoteBundle timerRemoteBundle =
+          stageBundleFactory.getBundle(receiverFactory, stateRequestHandler, progressHandler);
+      timerRemoteBundle.getInputReceivers().get(timerInputPCollection).accept(timerValue);
+      timerRemoteBundle.close();
+
+      timerData = stepContext.getNextFiredTimer(GlobalWindow.Coder.INSTANCE);
+    }
+  }
+
+  private void receive(String pCollectionId, Object receivedElement) throws Exception {
+    LOG.error("Received element {} for pcollection {}", receivedElement, pCollectionId);
+    // TODO(BEAM-6274): move this out into its own receiver class
+    if (timerOutputIdToSpecMap.containsKey(pCollectionId)) {
+      WindowedValue<KV<Object, Timer>> windowedValue =
+          (WindowedValue<KV<Object, Timer>>) receivedElement;
+      ProcessBundleDescriptors.TimerSpec timerSpec = timerOutputIdToSpecMap.get(pCollectionId);
+      Timer timer = windowedValue.getValue().getValue();
+
+      for (BoundedWindow window : windowedValue.getWindows()) {
+        Coder<BoundedWindow> windowCoder = timerWindowCodersMap.get(timerSpec.timerId());
+        StateNamespace namespace = StateNamespaces.window(windowCoder, window);
+
+        TimeDomain timeDomain = timerSpec.getTimerSpec().getTimeDomain();
+        String timerId = timerSpec.timerId();
+
+        DataflowExecutionContext.DataflowStepContext stepContext =
+            executionContext.getStepContext((DataflowOperationContext) this.context);
+
+        TimerInternals timerInternals = stepContext.namespacedToUser().timerInternals();
+        timerInternals.setTimer(namespace, timerId, timer.getTimestamp(), timeDomain);
+
+        timerIdToKey.put(timerId, windowedValue.getValue().getKey());
+        timerIdToPayload.put(timerId, timer.getPayload());
```

Comment:
Where is this data durably persisted?

Reply:
Does it need to be? In batch, if the worker crashes the whole work item will be reprocessed and the same timers will schedule again. In streaming, Windmill handles the persistence under the hood in their own TimerInternals.